### PR TITLE
Add directory for microbenchmarking

### PR DESCRIPTION
**Pull request by asaadaldien**
This needs definitely requires more iteration for documentation, improve scripts...etc
but for now IMO we can start with something simple for reproducibility  and iterate ?

### CODE REVIEW

File: iree/test/micro_benchmarks/mhlo_conv.mlir
Comment (MaheshRavishankar) [inline iree/test/micro_benchmarks/mhlo_conv.mlir:1]: Wondering if we can put a github code link here.

File: iree/test/micro_benchmarks/mhlo_conv.mlir
Comment (MaheshRavishankar) [inline iree/test/micro_benchmarks/mhlo_conv.mlir:26]: Code link on this one as well if possible?

File: iree/test/micro_benchmarks/mhlo_dot.mlir
Comment (MaheshRavishankar) [inline iree/test/micro_benchmarks/mhlo_dot.mlir:1]: Same as the conv ops, can we document where we are getting these from.

File: iree/test/micro_benchmarks/run_benchmark_on_android.sh
Comment (MaheshRavishankar) [inline iree/test/micro_benchmarks/run_benchmark_on_android.sh:6]: If we are checking this in, it needs more documentation.

**Review by MaheshRavishankar — CHANGES_REQUESTED** [review]
This would be good to have if you as me, but other people have a more informed opinion.

Only question though is could we just start from the `linalg.` operation? There are definitely advantages to using `mhlo`. Some ops might be lowered to a sequence of operations, but even then since we dont want anything `mhlo` in our core repo, starting from Linalg might be better.

File: iree/test/micro_benchmarks/run_benchmark_on_android.sh
Comment (benvanik) [reply iree/test/micro_benchmarks/run_benchmark_on_android.sh:6]: yeah, this would probably be better as a gist - otherwise it becomes yet another thing to support.

File: iree/test/micro_benchmarks/run_benchmark_on_android.sh
Comment (benvanik) [inline iree/test/micro_benchmarks/run_benchmark_on_android.sh:8]: vmfb :)

**Review by benvanik — COMMENTED** [review]
nit: "microbenchmark" is canonically one word so you can remove the underscore in the name (https://en.wiktionary.org/wiki/microbenchmark)

> Only question though is could we just start from the `linalg.` operation?

Agreed that linalg would be nice - soon we'll be removing HLO from IREE (relegating it only to the tensorflow importer) and as such none of these tests would be usable without doing a tensorflow importer build. That may be fine for these uses but then we may want to put these under microbenchmarks/mhlo/ so it's easy to exclude when someone is not building tf. If MHLO was the source of truth we could also have a script to run over the mhlo/ dir and generate a linalg dir, but I'm not sure how useful that is. We'll be having a lot of things coming from tosa too, and having a mix of linalg, mhlo, and tosa for microbenchmarking will be really hard to juggle as you'd need builds of all the different tools. That may still be worth it, though, if what you want to test includes the tosa/mhlo->linalg conversion - but if you think it's mostly just testing linalg then we should bake that out. I defer to you all and just don't want any surprises when I eventually remove the HLO passes from iree-opt/iree-translate :)

**Comment by asaadaldien** [comment]
> This would be good to have if you as me, but other people have a more informed opinion.

This has been my workflow for awhile iterating on performance on CPU.  

> Only question though is could we just start from the `linalg.` operation? There are definitely advantages to using `mhlo`. Some ops might be lowered to a sequence of operations, but even then since we dont want anything `mhlo` in our core repo, starting from Linalg might be better.


IMO having both are good for but specifically at this level (performance of an op or a sequence of ops (e.g fused conv-batchnorm) it easier to start with tosa or mhlo as it gets complicated adding all init_tensor, fill and constructing (or copy-paste) linalg.generics

File: iree/test/micro_benchmarks/mhlo_dot.mlir
Comment (asaadaldien) [reply iree/test/micro_benchmarks/mhlo_dot.mlir]: BTW, These ops also don't have to be sampled from a particular model (this coincidentally happened now as we are tracking e2e models performance and these ops are the slow path) they are mainly added to exercise  the codegen perf, e.g we can add an 3x4 conv to benchmark a winograd vs direct-conv implementation .

File: iree/test/micro_benchmarks/run_benchmark_on_android.sh
Comment (asaadaldien) [reply iree/test/micro_benchmarks/run_benchmark_on_android.sh]: The plan Is this should work on any linux-machine with local android phone attached =), 
I don't mind deleting but was to make it easy to reproduce the same numbers I have: 

had for this is users need to are:
1) Build an android release build
2) Define IREE_RELEASE_DIR and run the script.

File: iree/test/micro_benchmarks/run_benchmark_on_android.sh
Comment (hanhanW) [reply iree/test/micro_benchmarks/run_benchmark_on_android.sh]: We have something similar in python scripts.

https://github.com/google/iree/blob/main/build_tools/mako/compile_android_modules.py
https://github.com/google/iree/blob/main/build_tools/mako/benchmark_modules_on_android.py

The targets are configured in https://github.com/google/iree/blob/main/build_tools/mako/configuration.py

We can probably reuse them.

**Comment by hanhanW** [comment]
I think it depends on what we want to benchmark. Starting from Linalg could hide something like `linalg.fill`, or padding stuff. But yes, starting from Linalg is probably better because this is microbenchmark.

**Review by hanhanW — CHANGES_REQUESTED** [review]
I think it is meaningless if we only put files here. At least, we should translate them to IREE modules? And maybe use lit tests to check output?

I can help on connecting these to benchmarking infrastructure after we land it.

**Comment by asaadaldien** [comment]
> tarting from Linalg is probably better because this is microbenchmark.

Its very important to not drop fill, init_tensor, reshapes which is why doing something like matmul or various types of conv easy to be added here at the front end side (HLO or is easier otherwise it will get hard to reason about the performance of the op.

**Comment by asaadaldien** [comment]
> I think it is meaningless if we only put files here. At least, we should translate them to IREE modules? And maybe use lit tests to check output?
> 
> I can help on connecting these to benchmarking infrastructure after we land it.

Good ideas will be good to add rules to translate these and report the results on mako dashboard.
To separate concerns we shouldn't worry about correctness here as its covered in tests (if not then we add a test)

**Comment by MaheshRavishankar** [comment]
I have no opinions here. If this is deemed good to land, I'd be happy to use it.

**Comment by MaheshRavishankar** [comment]
> +1. Even if we convert to linalg ops, we should keep them as they are part of the original op. Otherwise we might measure something that is different. For example, right now the fill and copy generated from the original padded convolution op can be a large portion of the cost. Removing them would not reflect the current status.

I never said we should have a single linalg operation. You could have a sequence of Linalg operations (Which is more important). The goal is to eventually rid the IREE code base of MHLO ops.
How MHLO lowers to IREE or how TOSA lowers to IREE is immaterial. What we care about is that given a sequence of Linalg ops, whats the performance we get.

File: iree/test/micro_benchmarks/run_benchmark_on_android.sh
Comment (hanhanW) [reply iree/test/micro_benchmarks/run_benchmark_on_android.sh]: I remembered that I dropped comments about scripts somewhere, but it seems gone...

We have something similar in build_tools/mako/ which is mainly for mako benchmarks. We probably can reuse some of codes from there.

https://github.com/google/iree/blob/main/build_tools/mako/compile_android_modules.py
https://github.com/google/iree/blob/main/build_tools/mako/benchmark_modules_on_android.py

They are configured in https://github.com/google/iree/blob/main/build_tools/mako/configuration.py
I think we can add MICROBENCHMARKS, I'm happy to help on that once this gets check-in.

File: iree/test/microbenchmarks/mhlo_conv.mlir
Comment (hanhanW) [reply iree/test/microbenchmarks/mhlo_conv.mlir:3]: I think it worths to attach padding/stride informations in the function name, it can be showed on Mako benchmarks.

File: iree/test/microbenchmarks/mhlo_dot.mlir
Comment (hanhanW) [inline iree/test/microbenchmarks/mhlo_dot.mlir:44]: Add a new line

File: iree/test/microbenchmarks/mhlo_conv.mlir
Comment (hanhanW) [reply iree/test/microbenchmarks/mhlo_conv.mlir:3]: +1 on explain the naming scheme

**Comment by MaheshRavishankar** [comment]
The more I think about it, the more I dont like having operations lowered from mhlo for testing. I am not going to block this, but this will need to be deprecated/modified when we strip mhlo out of the core repo.
FYI @benvanik

File: iree/test/microbenchmarks/mhlo_conv.mlir
Comment (asaadaldien) [reply iree/test/microbenchmarks/mhlo_conv.mlir]: This is just an FYI the following are sampled from specific model (a peace extra information that is sometimes useful)

File: iree/test/microbenchmarks/mhlo_conv.mlir
Comment (asaadaldien) [reply iree/test/microbenchmarks/mhlo_conv.mlir:5]: The naming convention will get very long and they kind of implicit looking at the output size :)

File: iree/test/microbenchmarks/mhlo_dot_general.mlir
Comment (hanhanW) [inline iree/test/microbenchmarks/mhlo_dot_general.mlir:4]: Is the underline intended?

```suggestion
func @dot_general_4x384x32x384() -> tensor<4x384x384xf32> attributes { iree.module.export } {
```

**Review by hanhanW — APPROVED** [review]
Looks good to me. I'm not sure if others have more comment, but I approve for unblock merging it.

**Comment by asaadaldien** [comment]
@hanhanW In addition to mako tracking these for regression. IMO it will be very cool is to have a manual gitHub action users can trigger to run microbenchmarks on head vs this PR commit and post the textual as a comment on the PR. similar to what I usually do but not on my device/machine and more globally reproducible  :)

**Comment by hanhanW** [comment]
> @hanhanW In addition to mako tracking these for regression. IMO it will be very cool is to have a manual gitHub action users can trigger to run microbenchmarks on head vs this PR commit and post the textual as a comment on the PR. similar to what I usually do but not on my device/machine and more globally reproducible :)

Yes, we actually can do it on model benchmarks today. I usually do it manually through BuildKite. E.g., I'm now working on adding new metrics for linalg on tensors path, and I manually triggered a run based on current branch: https://buildkite.com/iree/iree-android-benchmark/builds/2195

We can probably integrate the action into label stuff (like kokoro:force_rerun). I think @GMNGeoffrey knows more about it.

**Pull request merged**
